spices/SPICE-0009-external-readers.adoc

```diff
@@ -0,0 +1,343 @@
+Custom readers help bridge Pkl into systems where it doesn't fit natively, enabling a wide variety of use cases:
+
+* Mediating access to resources that require authentication (eg. a secret store)
+* Querying data accessible via non-HTTP protocols (eg. LDAP)
+* Providing filesystem abstractions for remote or in-memory modules
+* And more!
+
+One of the primary drawbacks to custom readers is that using one instantly turns a Pkl module into a _incompatible language_: the module may only be evaluated by the client tool or application providing the reader and can no longer be evaluated directly using `pkl eval`.
+This hinders development and debugging of such modules, as typical workflows involving `trace(<expr>)` or `pkl eval -x '<expr>'` behave differently or are no longer possible.
+
+== Proposed Solution
+
+The existing message passing API and client libraries will be expanded to support external readers.
+Instead of client libraries invoking `pkl server`, `pkl eval` invocations, when configured with an external reader, will run the reader executable as a subprocess.
+External readers can provide any number of resource and/or module schemes and will be configured as an executable and list of arguments via:
+
+* CLI flag
+* PklProject property
+* Java API
+* Message passing API
+
+[source,pkl]
+----
+/// May be specified as an absolute path to an executable
+/// May also be specified as just an executable name, in which case it will be resolved according to the PATH environment variable
+executable: String
+
+/// Command line arguments that will be passed to the reader process
+arguments: Listing<String>
+----
+
+=== Example
+
+Consider this module:
+
+[source,pkl]
+----
+username = "john_appleseed"
+
+email = read("ldap://ds.example.com:389/dc=example,dc=com?mail?sub?(uid=\(username))").text
+----
+
+Pkl doesn't implement the `ldap:` resource URI scheme natively, but an external reader can provide it.
+Assuming a hypothetical `pkl-ldap` binary implementing the external reader protocol and the `ldap:` scheme is in the `$PATH`, this module can be evaluated as:
+
+[source,text]
+----
+$ pkl eval <module> --external-resource-reader ldap=pkl-ldap
+username = "john_appleseed"
+email = "appleseed@example.com"
+----
+
+In this example, the external reader may provide both `ldap:` and `ldaps:` schemes.
+External readers are registered by their scheme, so to support both schemes both would need to be specified on the command line:
+[source,text]
+----
+$ pkl eval <module> --external-resource-reader ldap=pkl-ldap --external-resource-reader ldaps=pkl-ldap
+----
+
+Registering an external reader for a scheme automatically adds that scheme to the default allowed modules/resources.
+As with Pkl's built-in module and resource schemes, setting explicit allowed module or resources overrides this behavior and appropriate patterns must be specified to allow use of external readers.
```

Comment:
Note on allowing registered external readers by default now that explicit scheme registration is needed.